pkg/controllers/hostedcluster/controller/controller.go

```diff
@@ -0,0 +1,215 @@
+package controller
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	hyperv1beta1 "github.com/openshift/hypershift/api/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	errors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes"
+	operatorclient "open-cluster-management.io/api/client/operator/clientset/versioned"
+	operatorapiv1 "open-cluster-management.io/api/operator/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	"github.com/stolostron/multicluster-controlplane/pkg/controllers/klusterlet/helpers"
+)
+
+const (
+	// hostedClusterImportAnnotation indicates whether to import hostedcluster as managed cluster
+	hostedClusterImportAnnotation = "cluster.open-cluster-management.io/import-managedcluster"
+	// hostedClusterImportFinalizer is the finalizer added to hostedcluster for importing as managed cluster
+	hostedClusterImportFinalizer = "cluster.open-cluster-management.io/import-managedcluster-finalizer"
+)
+
+// HostedClusterReconciler reconciles a HostedCluster object
+type HostedClusterReconciler struct {
+	client.Client
+	// Scheme *runtime.Scheme
+	ControlplaneKubeClient     kubernetes.Interface
+	ControlplaneOperatorClient operatorclient.Interface
+}
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+func (r *HostedClusterReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := log.FromContext(ctx)
+	log.V(1).Info("Reconciling hostecluster", "namespace", req.NamespacedName.Namespace, "name", req.NamespacedName.Name)
+
+	hostedCluster := &hyperv1beta1.HostedCluster{}
+	if err := r.Get(ctx, req.NamespacedName, hostedCluster); err != nil {
+		log.Error(err, "unable to fetch hostedcluster")
+		// we'll ignore not-found errors, since they can't be fixed by an immediate
+		// requeue (we'll need to wait for a new notification), and we can get them
+		// on deleted requests.
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	// hostedcluster is terminating, detach the managed cluster and return
+	if !hostedCluster.GetDeletionTimestamp().IsZero() {
+		if containStr(hostedCluster.GetFinalizers(), hostedClusterImportFinalizer) {
+			if err := r.pruneManagedClusterResources(ctx, log, hostedCluster); err != nil {
+				return ctrl.Result{}, err
+			}
+			hostedCluster.SetFinalizers(removeStr(hostedCluster.GetFinalizers(), hostedClusterImportFinalizer))
+			if err := r.Client.Update(ctx, hostedCluster, &client.UpdateOptions{}); err != nil {
```

Comment:
Would you use the patch instead of the update to update the finalizer, refer to
https://github.com/kubernetes-sigs/controller-runtime/issues/1509
https://github.com/kubernetes-sigs/controller-runtime/issues/741